src/main/cpp/src/datetime_rebase.cu

```diff
@@ -0,0 +1,213 @@
+        if (ymd >= gregorian_start) {
+          return d_input[idx];
+        }
+
+        // Reinterpret year/month/day as in Julian calendar then compute the days since epoch.
+        return cudf::timestamp_D{cudf::duration_D{days_from_julian(ymd)}};
+      });
+
+  return output;
+}
+
+/**
+ * @brief Struct store results of extracting time components from a timestamp.
+ */
+struct time_components {
+  int32_t hour;
+  int32_t minute;
+  int32_t second;
+  int32_t subsecond;
+};
+
+/**
+ * @brief Specialized modulo expression that handles negative values.
+ *
+ * @code{.pseudo}
+ * Examples:
+ *     modulo(1,60)  ->  1
+ *     modulo(-1,60) -> 59
+ * @endcode
+ */
+__device__ __inline__ auto modulo_time(int64_t time, int64_t base) {
+  return static_cast<int32_t>(((time % base) + base) % base);
+}
+
+/**
+ * @brief This function handles converting units by dividing and adjusting for negative values.
+ *
+ * @code{.pseudo}
+ * Examples:
+ *     scale(-61,60) -> -2
+ *     scale(-60,60) -> -1
+ *     scale(-59,60) -> -1
+ *     scale( 59,60) ->  0
+ *     scale( 60,60) ->  1
+ *     scale( 61,60) ->  1
+ * @endcode
+ */
+__device__ __inline__ int64_t scale_time(int64_t time, int64_t base) {
+  return (time - ((time < 0) * (base - 1L))) / base;
+}
+
+int64_t constexpr MICROS_PER_SECOND = 1'000'000L;
+
+__device__ __inline__ time_components get_time_components(int64_t micros) {
+  auto const subsecond = modulo_time(micros, MICROS_PER_SECOND);
+  micros = micros / MICROS_PER_SECOND - ((micros < 0) && (subsecond != 0));
+
+  auto const hour = modulo_time(scale_time(micros, 3600), 24);
+  auto const minute = modulo_time(scale_time(micros, 60), 60);
+  auto const second = modulo_time(micros, 60);
+
+  return time_components{hour, minute, second, subsecond};
+}
+
+// Convert the given number of microseconds since the epoch day 1970-01-01T00:00:00Z to a local
+// date-time in Proleptic Gregorian calendar, reinterpreting the result as in Julian calendar, then
+// compute the number of microseconds since the epoch from that Julian local date-time.
+// This is to match with Apache Spark's `rebaseGregorianToJulianMicros` function with timezone
+// fixed to UTC.
+std::unique_ptr<cudf::column> gregorian_to_julian_micros(cudf::column_view const &input,
+                                                         rmm::cuda_stream_view stream,
+                                                         rmm::mr::device_memory_resource *mr) {
+  auto output = cudf::make_timestamp_column(input.type(), input.size(),
+                                            cudf::detail::copy_bitmask(input, stream, mr),
+                                            input.null_count(), stream, mr);
+
+  thrust::transform(
+      rmm::exec_policy(stream), thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(input.size()),
+      output->mutable_view().begin<cudf::timestamp_us>(),
+      [d_input = input.begin<cudf::timestamp_us>()] __device__(auto const idx) {
```

Comment:
Do we want/need a check to be sure that the input type is micros?

Reply:
This is an internal function that is called only in the code path of micros timestamp. But yes, it doesn't hurt to add more check to prevent any accidents. I'll do it.